### PR TITLE
Feat: Add BitBucketRepository block

### LINKED ISSUE #2: Create `BitbucketRepository` block
Create a block that allows for interaction with a Bitbucket repository.

The block should have the following fields:
- `repository_url`: The URL of a Bitbucket repository.
- `reference`: An optional branch, tag, or commit to use when cloning the repository.
- `credentials`: An optional `BitbucketCredentials` block to use when cloning the repository. Necessary for cloning a private repository.

The block should implement the `ReadableDeploymentStorage` interface to allow it to be used a deployment storage.

### PR DESCRIPTION
**Pull request by serinamarie**
Closes #2 

Also fixes a test in test_credentials.py that was misleading ([ln 21](https://github.com/PrefectHQ/prefect-bitbucket/compare/main...feat/repository_block?expand=1#diff-36141597d563fa86c916ffba13f2c1c3b247a89ef19f5f297256aac11867f393R21)).
Added pydocstyle.

### Screenshots
Screen Shot 2022-11-29 at 9 40 22 AM

Screen Shot 2022-11-29 at 10 44 39 AM

### Checklist

Screen Shot 2022-11-29 at 4 05 49 PM

- [x] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-bitbucket/issues/new/choose) first.
- [x] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [x] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-bitbucket/blob/main/CHANGELOG.md)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - `BitBucketCredentials` block - [#3](https://github.com/PrefectHQ/prefect-bitbucket/pull/3/)
+- `BitBucketRepository` block - [#5](https://github.com/PrefectHQ/prefect-bitbucket/pull/4/)
```
Comment (ahuang11) [inline CHANGELOG.md]:
```suggestion
- `BitBucketRepository` block - [#5](https://github.com/PrefectHQ/prefect-bitbucket/pull/5/)
```

File: prefect_bitbucket/repository.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        description="The URL of a BitBucket repository to read from in HTTPS format",
+    )
+    reference: Optional[str] = Field(
+        default=None,
+        description="An optional reference to pin to; can be a branch or tag.",
+    )
+    credentials: Optional[BitBucketCredentials] = Field(
+        default=None,
+        description="An optional BitBucket Credentials block for authenticating with "
+        "private BitBucket repos.",
```
Comment (ahuang11) [inline prefect_bitbucket/repository.py]:
```suggestion
        description=(
            "An optional BitBucket Credentials block for authenticating with "
            "private BitBucket repos."
        ),
```

File: README.md [hunk 3]
```diff
@@ -81,6 +79,80 @@ def create_new_bitbucket_creds_flow():
...
+    username="my-username"  # optional
+)
+
+# Saves the BitBucketCredentials block to your Prefect workspace (in the Blocks tab)
+bitbucket_credentials_block.save(name="my-bitbucket-credentials-block")
+
+private_repo = "https://bitbucket.org/my-workspace/my-repository.git"
+
+# Creates a private BitBucket repository BitBucketRepository block
+private_bitbucket_block = BitBucketRepository(
```
Comment (ahuang11) [inline README.md]:
```suggestion

# Creates a private BitBucket repository BitBucketRepository block
private_repo = "https://bitbucket.org/my-workspace/my-repository.git"
private_bitbucket_block = BitBucketRepository(
```

File: README.md [hunk 4]
```diff
@@ -81,6 +79,80 @@ def create_new_bitbucket_creds_flow():
...
+```python
+from prefect_bitbucket import BitBucketCredentials, BitBucketRepository
+
+# Loads a preexisting BitBucketCredentials block
+bitbucket_credentials_block.load(name="my-bitbucket-credentials-block")
+
+private_repo = "https://bitbucket.org/my-workspace/my-repository.git"
+
+# Creates a private BitBucket repository BitBucketRepository block
+private_bitbucket_block = BitBucketRepository(
```
Comment (ahuang11) [inline README.md]:
```suggestion

# Creates a private BitBucket repository BitBucketRepository block
private_repo = "https://bitbucket.org/my-workspace/my-repository.git"
private_bitbucket_block = BitBucketRepository(
```

File: tests/test_repository.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+from prefect.exceptions import InvalidRepositoryURLError
+from prefect.testing.utilities import AsyncMock
+from pydantic import SecretStr
+
+import prefect_bitbucket
+from prefect_bitbucket.credentials import BitBucketCredentials
+from prefect_bitbucket.repository import BitBucketRepository
+
+
+class TestBitBucket:
```
Comment (ahuang11) [inline tests/test_repository.py]:
```suggestion
class TestBitBucketRepository:
```

**Review by ahuang11 — COMMENTED** [review]
Thanks for adding all the examples and piloting pydocstyle!

I am wondering whether setup.cfg is ignoring too many codes, especially the D1s D2s and D3s http://www.pydocstyle.org/en/2.1.1/error_codes.html

`ignore = D100,D101,D104,D107,D212,D200,D202,D213,D203,D205,D300,D400,D401,D405,D406,D407,D415`

File: prefect_bitbucket/repository.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                    (
+                        "Credentials can only be used with BitBucket repositories "
+                        "using the 'HTTPS' format. You must either remove the "
+                        "credential if you wish to use the 'SSH' format and are not "
+                        "using a private repository, or you must change the repository "
+                        "URL to the 'HTTPS' format."
+                    )
+                )
+
+        return v
```
Comment (ahuang11) [inline prefect_bitbucket/repository.py]: Can we add a test where `v is None`?

**Comment by serinamarie** [comment]
> Thanks for adding all the examples and piloting pydocstyle!
> 
> I am wondering whether setup.cfg is ignoring too many codes, especially the D1s D2s and D3s http://www.pydocstyle.org/en/2.1.1/error_codes.html
> 
> `ignore = D100,D101,D104,D107,D212,D200,D202,D213,D203,D205,D300,D400,D401,D405,D406,D407,D415`

Hi Andrew, I pretty much ignored the ones that interfered with `black` or were completely opposite to how the other repos' style is.

**Comment by ahuang11** [comment]
> > Thanks for adding all the examples and piloting pydocstyle!
> > I am wondering whether setup.cfg is ignoring too many codes, especially the D1s D2s and D3s http://www.pydocstyle.org/en/2.1.1/error_codes.html
> > `ignore = D100,D101,D104,D107,D212,D200,D202,D213,D203,D205,D300,D400,D401,D405,D406,D407,D415`
> 
> Hi Andrew, I pretty much ignored the ones that interfered with `black` or were completely opposite to how the other repos' style is.

Thanks for testing it. I am wondering if you include D100 and D101, which ones will it complain about?

File: prefect_bitbucket/repository.py (hunk 6 above)
Comment (serinamarie) [reply prefect_bitbucket/repository.py]: v shouldn't be None since we are only running it if a credential is provided. What would be ensuring in that case, since there would be no credentials to compare the scheme with?

Comment (ahuang11) [reply prefect_bitbucket/repository.py]: Sorry, I meant a unit test that passes through this validator when `credentials=None`, e.g. instantiating BitBucketRepository without credentials set.

File: prefect_bitbucket/repository.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+    repository: str = Field(
+        default=...,
+        description="The URL of a BitBucket repository to read from in HTTPS format",
+    )
+    reference: Optional[str] = Field(
+        default=None,
+        description="An optional reference to pin to; can be a branch or tag.",
+    )
+    credentials: Optional[BitBucketCredentials] = Field(
```
Comment (ahuang11) [inline prefect_bitbucket/repository.py]: I am slightly leaning towards renaming this to `bitbucket_credentials` even though it's verbose since all our other Collections prefixes `credentials` with the collection name and it's clearer to the user that only "BitBucketCredentials" is allowed.

File: README.md [hunk 8]
```diff
@@ -81,6 +79,80 @@ def create_new_bitbucket_creds_flow():
 create_new_bitbucket_creds_flow()
 ```
 
+#### Create a BitBucketRepository block for a public repo.
```
Comment (ahuang11) [inline README.md]: Would you mind sharing how the documentation right sidebar looks with a period in the section header?

**Comment by serinamarie** [comment]
> D100,D101
```
tests/test_repository.py:16 in public class `TestBitBucketRepository`: D101: Missing docstring in public class
setup.py:1 at module level: D100: Missing docstring in public module
tests/test_repository.py:1 at module level: D100: Missing docstring in public module
tests/conftest.py:1 at module level: D100: Missing docstring in public module
tests/test_credentials.py:1 at module level: D100: Missing docstring in public module
```

**Comment by ahuang11** [comment]
Thanks! Perhaps we can just ignore `setup.py` and `tests/` rather than the codes themselves: https://github.com/PyCQA/pydocstyle/issues/186#issuecomment-631760965

File: prefect_bitbucket/repository.py (hunk 7 above)
Comment (serinamarie) [reply prefect_bitbucket/repository.py]: Done

File: README.md (hunk 8 above)
Comment (serinamarie) [reply README.md]: For sure

Screen Shot 2022-11-29 at 2 55 55 PM

Comment (ahuang11) [reply README.md]: I guess the periods look fine, but if we do use periods, maybe Write and run a flow can use one too.

Comment (serinamarie) [reply README.md]: @ahuang11 I personally think no period looks better, but what do you think?

Comment (ahuang11) [reply README.md]: I don't have a strong preference, but if we go with precedent in other collections, then no periods.

**Comment by ahuang11** [comment]
Can we also drop the other D1s and D2s and D3s codes from being ignored as well? D4s, maybe some of them can be useful.

File: prefect_bitbucket/credentials.py [hunk 9]
```diff
@@ -7,20 +7,23 @@
...
 
 class BitBucketCredentials(Block):
-    """
-    Store a BitBucket access token to interact with private BitBucket
-    repositories.
+    """Store BitBucket credentials to interact with private BitBucket repositories.
 
-    Attributes:
+    Attributes
+    ----------
```
Comment (ahuang11) [inline prefect_bitbucket/credentials.py]: Really thank you for trying out pydocstyle! Can we disable the pydocstyle code that requires this; I think I prefer the original `Attributes:`

File: setup.cfg [hunk 10]
```diff
@@ -40,5 +40,5 @@ asyncio_mode = auto
 
 [pydocstyle]
 inherit = false
-ignore = D200,D202,D213,D203,D205,D300,D400,D401,D405,D406,D407,D415
+ignore = D213,D203  # D203 and D211 are in conflict, D212 and D213 are in conflict
```
Comment (ahuang11) [inline setup.cfg]: Thanks for testing! For more info: https://github.com/PyCQA/pydocstyle/issues/475#issuecomment-627520864

File: prefect_bitbucket/repository.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    # private BitBucket repository
+    private_bitbucket_block = BitBucketRepository(
+        repository="https://bitbucket.com/my-project/my-repository.git",
+        bitbucket_credentials=BitBucketCredentials.load("my-bitbucket-credentials-block")
+
+
+    )
+
+    private_bitbucket_block.save(name="my-private-bitbucket-block")
+```
```
Comment (ahuang11) [inline prefect_bitbucket/repository.py]: I think this needs unindent.
image

```suggestion

Examples
```python
from prefect_bitbucket.repository import BitBucketRepository

# public BitBucket repository
public_bitbucket_block = BitBucketRepository(
    repository="https://bitbucket.com/my-project/my-repository.git"
)

public_bitbucket_block.save(name="my-bitbucket-block")

# specific branch or tag
branch_bitbucket_block = BitBucketRepository(
    reference="branch-or-tag-name",
    repository="https://bitbucket.com/my-project/my-repository.git"
)

branch_bitbucket_block.save(name="my-bitbucket-block")

# private BitBucket repository
private_bitbucket_block = BitBucketRepository(
    repository="https://bitbucket.com/my-project/my-repository.git",
    bitbucket_credentials=BitBucketCredentials.load("my-bitbucket-credentials-block")
)

private_bitbucket_block.save(name="my-private-bitbucket-block")
```

File: prefect_bitbucket/repository.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    async def get_directory(
+        self, from_path: Optional[str] = None, local_path: Optional[str] = None
+    ) -> None:
+        """Clones a BitBucket project specified in `from_path` to the provided `local_path`.
+
+        This defaults to cloning the repository reference configured on the
+        Block to the present working directory.
+
+        Args:
+        ----
```
Comment (ahuang11) [inline prefect_bitbucket/repository.py]:
```suggestion
```

File: tests/test_repository.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+        Returns a tuple with the expected parent-level contents and the
+        expected child-level contents.
+        """
+
+        # add file to tmp_src
+        f1_name = "dog.text"
+        f1_path = Path(tmp_src) / f1_name
+        f1 = open(f1_path, "w")
+        f1.close()
```
Comment (ahuang11) [inline tests/test_repository.py:30]: This is fine, but I think you can also do
`f1_path.write_text("")`

**Review by ahuang11 — APPROVED** [review]
Thanks! It looks good to me.
